### PR TITLE
Make catch clause checking consistent with variable declarations

### LINKED ISSUE #47383: Destructuring of `unknown` catch variable is not an error
# Bug Report

### 🔎 Search Terms

catch, destructure, destructuring, useUnknownInCatchVariables

### 🕗 Version & Regression Information

- This is the behavior in every version I tried: v4.0.5, v4.5.4

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.5.4#code/DYUwLgBAhgXBCuA7A1og9gd0QbgFCkgG8QAnEtEgXwgF5o9cAzJAYzAEs1EIAjKEgBSFGaNJThJUmRAEoIhSrlxgSAT3m5qLKGBYALCENLkqElOixzCmoA)

### 💻 Code

```ts
let a: unknown;
let {foo} = a;
//   ^^^ Property 'foo' does not exist on type 'unknown'.(2339)

function bar({foo}: unknown) {}
//            ^^^ Property 'foo' does not exist on type 'unknown'.(2339)

try {
} catch ({foo}: unknown) {
  // foo is type any
  // with useUnknownInCatchVariables foo is type unknown
}
```

### 🙁 Actual behavior

Destructuring `unknown` is an error, but destructuring a catch variable, which is of type `unknown` is not an error.

### 🙂 Expected behavior

I would expect destructuring of a catch variable of type `unknown` to produce an error.

### PR DESCRIPTION
**Pull request by jakebailey**
Fixes #47383

This changes our handling of catch clauses to work just like other variable declarations (by using the same code!), with the exception that:

- You can only annotate the variable declaration with `any` or `unknown`. This is the same as before.
- If you don't annotate it, we treat it as an implicit annotation of `any` or `unknown` depending on `useUnknownInCatchVariables`.
	- This is new; previously, the code for catch clause typing was in a totally different place which caused all destructured values to have the same type depending on that setting (very different than regular variable declarations).
	- We treat it as having an explicit annotation because if we choose `any`, we don't want to see follow-on errors about implicit any within the destructured value. (Or, maybe we do? Probably not?)

Now that we check these declarations using the same code as variable declarations, we can correctly and consistently handle destructuring within catch clauses, fixing the linked issue. It also seems to fix issues I've observed in tsserver as the old error was only surfaced as a side effect (so, sometimes errors would appear thanks to semantic tokens and such).

This is a breaking change as you'll get errors on destructure if the type is unknown, but I think it's more consistent.

### CODE REVIEW

**Comment by jakebailey** [comment]
@typescript-bot test this
@typescript-bot test top100
@typescript-bot user test this
@typescript-bot user test tsserver
@typescript-bot test tsserver top100
@typescript-bot run dt
@typescript-bot perf test this
@typescript-bot pack this

**Comment by jakebailey** [comment]
Also, with his change in place, it would be really, really easy to allow arbitrary type annotations on catch clauses.

File: tests/baselines/reference/destructureCatchClause(strict=false,useunknownincatchvariables=true).errors.txt [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+tests/cases/compiler/destructureCatchClause.ts(32,15): error TS2461: Type 'unknown' is not an array type.
+tests/cases/compiler/destructureCatchClause.ts(33,15): error TS2461: Type 'unknown' is not an array type.
+tests/cases/compiler/destructureCatchClause.ts(35,17): error TS2339: Property 'a' does not exist on type '{}'.
+
+
+==== tests/cases/compiler/destructureCatchClause.ts (14 errors) ====
+    // These are okay with useUnknownInCatchVariables=false, but not okay with useUnknownInCatchVariables=true.
+    try {} catch ({ x }) { x }
+                    ~
+!!! error TS2339: Property 'x' does not exist on type '{}'.
```
Comment (jakebailey) [inline tests/baselines/reference/destructureCatchClause(strict=false,useunknownincatchvariables=true).errors.txt:21]: These kinds of errors happen in `strictNullChecks=false` because destructuring an object uses the apparent type, and the apparent type of `unknown` in `strictNullChecks=false` is `{}`. See #47494.

I can't say I like these errors (especially given the errors for destructuring an array actually _do_ mention unknown), but I feel like the likelihood of someone enabling `useUnknownInCatchVariables` but not enabling `strictNullChecks` is very low.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -27079,6 +27078,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             isFunctionLike(node) && !getImmediatelyInvokedFunctionExpression(node) ||
             node.kind === SyntaxKind.ModuleBlock ||
             node.kind === SyntaxKind.SourceFile ||
+            node.kind === SyntaxKind.CatchClause ||
```
Comment (sandersn) [inline src/compiler/checker.ts:27081]: why is this needed? How much does it increase the tree-walking that control flow analysis does?

**Review by sandersn — APPROVED** [review]
Look good, although I don't know what the change to control flow does.

File: src/compiler/checker.ts (hunk 2 above)
Comment (jakebailey) [reply src/compiler/checker.ts:27081]: Right now, this code is wrong for catch clauses because it doesn't stop at the CatchClause block (where the catch clause variable declaration is scoped) and goes to the try/catch's enclosing scope.

Then, it determines that the catch clause variable is not yet defined (because it's checking the wrong scope!), but then because we only allow `any` or `unknown`, `assumeInitialized` turns into `true`, saving the day.

I noticed this on the PR where I allowed _any_ annotation because this code path happens to be called by semantic tokens and causes errors that persist; I cherry-picked it even though it's not strictly required because 1) it's definitely wrong, and 2) it's faster to skip all of the work if we know it's a catch clause variable.

Comment (jakebailey) [reply src/compiler/checker.ts:27081]: (Half-related, but I swear I read code during working on this PR that looked the same as this code but was correct; I can't remember where anymore...)

Comment (sandersn) [reply src/compiler/checker.ts:27081]: faster is good 👍🏼

Comment (jakebailey) [reply src/compiler/checker.ts:27081]: I think it could be faster, even; assumeInitialized is a huge statement and I think a bit of work could be skipped if some of the more simple conditions were true.
